openmdao/core/system.py

```diff
@@ -3372,6 +3372,16 @@ def get_io_metadata(self, iotypes=('input', 'output'), metadata_keys=None,
         prefix = self.pathname + '.' if self.pathname else ''
         rel_idx = len(prefix)
 
+        # Setup any modified subsystems so the metadata dicts are up-to-date.
+        if self._problem_meta:
+            conf_info = self._problem_meta['config_info']
+            if conf_info:
+                if self._subsystems_allprocs:
+                    for s in conf_info._modified_system_iter(self):
+                        s._setup_var_data()
+                    conf_info._reset()
```

Comment:
Actually I would move the _reset() call to after the self._setup_var_data, and would put the self._setup_var_data inside a `if config_info._modified_systems:` conditional so it won't get called unnecessarily.